File: src/ui/settingsdialog.cpp
#include "settingsdialog.h"
#include "components/arrowspinbox.h"
#include "../core/algorithms/gs_algorithm.h"

#include <QCheckBox>
#include <QDialogButtonBox>
#include <QFormLayout>
#include <QGroupBox>
#include <QLabel>
#include <QLineEdit>
#include <QPushButton>
#include <QStandardItemModel>
#include <QStringList>
#include <QVBoxLayout>
#include <QScreen>
#include <QGuiApplication>
#include <QtGlobal>
#include <QTabWidget>

SettingsDialog::SettingsDialog(int slmW, int slmH, double slmPix, int backend,
                               int camW, int camH, double camPix,
                               const QString &udpBindIp, int udpPort,
                               double wave, double focal, double cameraImagingMagnification, int slmOutputMode,
                               bool autoRunGsEnabled,
                               bool autoSendSlmEnabled,
                               int startingPhaseMaskMode,
                               int gsComputeBackendMode,
                               int openClPlatformIndex,
                               int openClDeviceIndex,
                               int cudaDeviceIndex,
                               int cameraRotation, bool flipX, bool saveCompressed,
                               bool saveFollowsTransforms, bool flipY,
                               QWidget *parent)
    : QDialog(parent) {

    setWindowTitle("Hardware Settings");
    setMinimumWidth(450);

    // Resize to fit primary screen height with a small buffer
    QScreen *screen = QGuiApplication::primaryScreen();
    if (screen) {
        const int availableHeight = screen->availableGeometry().height();
        resize(500, availableHeight - 60);
    } else {
        resize(500, 800);
    }

    QVBoxLayout *mainLayout = new QVBoxLayout(this);
    mainLayout->setContentsMargins(12, 12, 12, 12);
    mainLayout->setSpacing(8);

    QTabWidget *tabWidget = new QTabWidget(this);
    
    // --- HARDWARE TAB ---
    QWidget *hardwareTab = new QWidget();
    QVBoxLayout *hardwareLayout = new QVBoxLayout(hardwareTab);
    hardwareLayout->setContentsMargins(12, 12, 12, 12);
    hardwareLayout->setSpacing(8);

    // --- 1. SLM SETTINGS ---
    QGroupBox *slmGroup = new QGroupBox("SLM Parameters");
    QFormLayout *slmForm = new QFormLayout();

    widthSpin = new ArrowSpinBox(this); widthSpin->setRange(100, 8000); widthSpin->setValue(slmW);
    heightSpin = new ArrowSpinBox(this); heightSpin->setRange(100, 8000); heightSpin->setValue(slmH);
    pixelSpin = new ArrowDoubleSpinBox(this); pixelSpin->setRange(0.1, 100.0); pixelSpin->setDecimals(2);
    pixelSpin->setSuffix(" um"); pixelSpin->setValue(slmPix);

    slmOutputModeCombo = new QComboBox(this);
    slmOutputModeCombo->addItems({"DLL", "Direct Screen"});
    slmOutputModeCombo->setCurrentIndex((slmOutputMode == 1) ? 1 : 0);

    autoRunGsCheck = new QCheckBox(this);
    autoRunGsCheck->setChecked(autoRunGsEnabled);

    autoSendSlmCheck = new QCheckBox(this);
    autoSendSlmCheck->setChecked(autoSendSlmEnabled);

    startingPhaseMaskCombo = new QComboBox(this);
    startingPhaseMaskCombo->addItem("Binary Grating", 1);
    startingPhaseMaskCombo->addItem("Checkerboard Pattern", 0);
    startingPhaseMaskCombo->addItem("Random Phase", 2);
    const int safeMode = qBound(0, startingPhaseMaskMode, 2);
    const int safeIndex = startingPhaseMaskCombo->findData(safeMode);
    startingPhaseMaskCombo->setCurrentIndex((safeIndex >= 0) ? safeIndex : 0);

    computeBackendCombo = new QComboBox(this);
#if HOT_ENABLE_CUDA_GS
    computeBackendCombo->addItem("Auto (CUDA -> OpenCL -> CPU)", 0);
#else
    computeBackendCombo->addItem("Auto (OpenCL -> CPU)", 0);
#endif
    computeBackendCombo->addItem("CPU", 1);
    computeBackendCombo->addItem("OpenCL", 2);
    computeBackendCombo->addItem("CUDA", 3);

#if !HOT_ENABLE_CUDA_GS
    const int cudaBackendRow = computeBackendCombo->findData(3);
    if (cudaBackendRow >= 0) {
        QStandardItemModel *model = qobject_cast<QStandardItemModel *>(computeBackendCombo->model());
        if (model) {
            QStandardItem *item = model->item(cudaBackendRow);
            if (item) {
                item->setText("CUDA (unavailable in this build)");
                item->setEnabled(false);
            }
        }
    }
#endif

    int requestedBackendMode = gsComputeBackendMode;
#if !HOT_ENABLE_CUDA_GS
    if (requestedBackendMode == 3) {
        requestedBackendMode = 0;
    }
#endif
    const int safeBackendIndex = computeBackendCombo->findData(requestedBackendMode);
    computeBackendCombo->setCurrentIndex((safeBackendIndex >= 0) ? safeBackendIndex : 0);

    openClDeviceCombo = new QComboBox(this);
    const QVector<GSAlgorithm::GSOpenClDeviceInfo> openClDevices = GSAlgorithm::enumerateOpenClDevices();
    if (openClDevices.isEmpty()) {
        openClDeviceCombo->addItem("No OpenCL devices detected", "-1:-1");
    } else {
        for (const GSAlgorithm::GSOpenClDeviceInfo &device : openClDevices) {
            const QString key = QString("%1:%2").arg(device.platformIndex).arg(device.deviceIndex);
            openClDeviceCombo->addItem(device.displayName, key);
        }
    }

    const QString targetDeviceKey = QString("%1:%2").arg(openClPlatformIndex).arg(openClDeviceIndex);
    const int selectedDeviceIdx = openClDeviceCombo->findData(targetDeviceKey);
    if (selectedDeviceIdx >= 0) {
        openClDeviceCombo->setCurrentIndex(selectedDeviceIdx);
    } else {
        openClDeviceCombo->setCurrentIndex(0);
    }

    cudaDeviceCombo = new QComboBox(this);
#if HOT_ENABLE_CUDA_GS
    const QVector<GSAlgorithm::GSCudaDeviceInfo> cudaDevices = GSAlgorithm::enumerateCudaDevices();
    if (cudaDevices.isEmpty()) {
        cudaDeviceCombo->addItem("No CUDA devices detected", -1);
    } else {
        for (const GSAlgorithm::GSCudaDeviceInfo &device : cudaDevices) {
            QString display = device.displayName;
            if (!device.isCompatible) {
                display += " [Incompatible]";
            }
            cudaDeviceCombo->addItem(display, device.deviceIndex);
        }
    }
#else
    cudaDeviceCombo->addItem("CUDA backend unavailable in this build", -1);
#endif

    const int cudaSelectedIdx = cudaDeviceCombo->findData(cudaDeviceIndex);
    if (cudaSelectedIdx >= 0) {
        cudaDeviceCombo->setCurrentIndex(cudaSelectedIdx);
    } else {
        cudaDeviceCombo->setCurrentIndex(0);
    }

    slmForm->addRow("SLM Width (pixels):", widthSpin);
    slmForm->addRow("SLM Height (pixels):", heightSpin);
    slmForm->addRow("SLM Pixel Size:", pixelSpin);
    slmForm->addRow("SLM Output Mode:", slmOutputModeCombo);
    slmForm->addRow("Auto-run GS:", autoRunGsCheck);
    slmForm->addRow("Auto-send SLM:", autoSendSlmCheck);
    slmForm->addRow("Starting Phase Mask:", startingPhaseMaskCombo);
    slmForm->addRow("GS Compute Backend:", computeBackendCombo);
    slmForm->addRow("OpenCL Device:", openClDeviceCombo);
    slmForm->addRow("CUDA Device:", cudaDeviceCombo);
    slmGroup->setLayout(slmForm);

    auto refreshDeviceEnabledState = [this]() {
        const int backendMode = computeBackendCombo->currentData().toInt();
        const bool backendNeedsOpenCl = (backendMode == 0 || backendMode == 2);
        const bool backendNeedsCuda = (backendMode == 0 || backendMode == 3);
        const bool hasValidDevice = openClDeviceCombo->currentData().toString() != "-1:-1";
        const bool hasValidCudaDevice = cudaDeviceCombo->currentData().toInt() >= 0;
        openClDeviceCombo->setEnabled(backendNeedsOpenCl && hasValidDevice);
        cudaDeviceCombo->setEnabled(backendNeedsCuda && hasValidCudaDevice);
    };
    connect(computeBackendCombo,
            QOverload<int>::of(&QComboBox::currentIndexChanged),
            this,
            [refreshDeviceEnabledState](int index) {
                Q_UNUSED(index);
                refreshDeviceEnabledState();
            });
    refreshDeviceEnabledState();

    // --- 2. CAMERA SETTINGS ---
    QGroupBox *camGroup = new QGroupBox("Camera Parameters");
    QFormLayout *camForm = new QFormLayout();

    cameraBackendCombo = new QComboBox(this);
    cameraBackendCombo->addItems({"Qt Native (WMF)", "OpenCV (DirectShow)", "UDP Stream (Ethernet)"});
    cameraBackendCombo->setCurrentIndex(qBound(0, backend, 2));

    camWidthSpin = new ArrowSpinBox(this); camWidthSpin->setRange(100, 8000); camWidthSpin->setValue(camW);
    camHeightSpin = new ArrowSpinBox(this); camHeightSpin->setRange(100, 8000); camHeightSpin->setValue(camH);
    camPixelSpin = new ArrowDoubleSpinBox(this); camPixelSpin->setRange(0.1, 100.0); camPixelSpin->setDecimals(2);
    camPixelSpin->setSuffix(" um"); camPixelSpin->setValue(camPix);
    udpBindIpEdit = new QLineEdit(this);
    udpBindIpEdit->setPlaceholderText("0.0.0.0");
    udpBindIpEdit->setText(udpBindIp.isEmpty() ? QStringLiteral("0.0.0.0") : udpBindIp);
    udpPortSpin = new ArrowSpinBox(this); udpPortSpin->setRange(1, 65535); udpPortSpin->setValue(qBound(1, udpPort, 65535));

    camForm->addRow("Camera Engine:", cameraBackendCombo);
    camForm->addRow("Camera Width (px):", camWidthSpin);
    camForm->addRow("Camera Height (px):", camHeightSpin);
    camForm->addRow("Camera Pixel Size:", camPixelSpin);
    camGroup->setLayout(camForm);

    // --- 3. OPTICAL SETUP ---
    QGroupBox *opticsGroup = new QGroupBox("Optical Setup");
    QFormLayout *opticsForm = new QFormLayout();

    waveSpin = new ArrowDoubleSpinBox(this); waveSpin->setRange(100.0, 2000.0); waveSpin->setDecimals(1);
    waveSpin->setSuffix(" nm"); waveSpin->setValue(wave);

    focalSpin = new ArrowDoubleSpinBox(this); focalSpin->setRange(1.0, 1000.0); focalSpin->setDecimals(1);
    focalSpin->setSuffix(" mm"); focalSpin->setValue(focal);

    cameraMagnificationSpin = new ArrowDoubleSpinBox(this);
    cameraMagnificationSpin->setRange(0.01, 1000.0);
    cameraMagnificationSpin->setDecimals(3);
    cameraMagnificationSpin->setValue(qMax(0.01, cameraImagingMagnification));
    cameraMagnificationSpin->setToolTip("Magnification between the trap plane and the camera sensor. "
                                        "Use this to calibrate spot placement without changing the real Fourier lens focal length.");

    opticsForm->addRow("Laser Wavelength:", waveSpin);
    opticsForm->addRow("Fourier Lens Focal Length (physical):", focalSpin);
    opticsForm->addRow("Camera Imaging Magnification:", cameraMagnificationSpin);
    opticsGroup->setLayout(opticsForm);

    QLabel *monitorHint = new QLabel("Monitor target is selected from Tools > Select Monitor.");
    monitorHint->setWordWrap(true);

    // Assemble the hardware tab
    hardwareLayout->addWidget(slmGroup);
    hardwareLayout->addWidget(camGroup);
    hardwareLayout->addWidget(opticsGroup);
    hardwareLayout->addWidget(monitorHint);
    hardwareLayout->addStretch();

    // --- UI SETTINGS TAB ---
    QWidget *uiTab = new QWidget();
    QFormLayout *uiLayout = new QFormLayout(uiTab);
    uiLayout->setContentsMargins(12, 12, 12, 12);
    uiLayout->setSpacing(8);

    cameraRotationCombo = new QComboBox(this);
    cameraRotationCombo->addItems({"0", "90", "180", "270"});
    int rotIndex = 0;
    if (cameraRotation == 90) rotIndex = 1;
    else if (cameraRotation == 180) rotIndex = 2;
    else if (cameraRotation == 270) rotIndex = 3;
    cameraRotationCombo->setCurrentIndex(rotIndex);

    flipXCheck = new QCheckBox(this);
    flipXCheck->setChecked(flipX);

    flipYCheck = new QCheckBox(this);
    flipYCheck->setChecked(flipY);

    saveCompressedCheck = new QCheckBox(this);
    saveCompressedCheck->setChecked(saveCompressed);

    saveFollowsTransformsCheck = new QCheckBox(this);
    saveFollowsTransformsCheck->setChecked(saveFollowsTransforms);

    QLabel *udpHeader = new QLabel("<b>UDP Stream Settings</b>", this);
    QLabel *captureHeader = new QLabel("<b>Capture Save Settings</b>", this);

    uiLayout->addRow(udpHeader);
    uiLayout->addRow("UDP Bind IP:", udpBindIpEdit);
    uiLayout->addRow("UDP Port:", udpPortSpin);
    uiLayout->addRow(captureHeader);
    uiLayout->addRow("Camera Rotation (deg):", cameraRotationCombo);
    uiLayout->addRow("Flip Camera X:", flipXCheck);
    uiLayout->addRow("Flip Camera Y:", flipYCheck);
    uiLayout->addRow("Save Compressed:", saveCompressedCheck);
    uiLayout->addRow("Save Using Rotation/Flip:", saveFollowsTransformsCheck);

    // Add Tabs
    tabWidget->addTab(hardwareTab, "Hardware");
    tabWidget->addTab(uiTab, "UI Settings");

    mainLayout->addWidget(tabWidget);

    QDialogButtonBox *buttonBox = new QDialogButtonBox(QDialogButtonBox::Save | QDialogButtonBox::Apply | QDialogButtonBox::Cancel, Qt::Horizontal, this);
    mainLayout->addWidget(buttonBox);

    QPushButton *applyBtn = buttonBox->button(QDialogButtonBox::Apply);
    if (applyBtn) {
        connect(applyBtn, &QPushButton::clicked, this, &SettingsDialog::applyRequested);
    }

    connect(buttonBox, &QDialogButtonBox::accepted, this, &QDialog::accept);
    connect(buttonBox, &QDialogButtonBox::rejected, this, &QDialog::reject);
}

// UI Settings Getters
int SettingsDialog::getCameraRotation() const { return cameraRotationCombo->currentText().toInt(); }
bool SettingsDialog::getFlipX() const { return flipXCheck->isChecked(); }
bool SettingsDialog::getFlipY() const { return flipYCheck->isChecked(); }
bool SettingsDialog::getSaveCompressed() const { return saveCompressedCheck->isChecked(); }
bool SettingsDialog::getSaveFollowsTransforms() const { return saveFollowsTransformsCheck->isChecked(); }

// SLM Getters
int SettingsDialog::getWidth() const { return widthSpin->value(); }
int SettingsDialog::getHeight() const { return heightSpin->value(); }
double SettingsDialog::getPixelSize() const { return pixelSpin->value(); }

// Camera Getters
int SettingsDialog::getCameraBackend() const { return cameraBackendCombo->currentIndex(); }
int SettingsDialog::getCamWidth() const { return camWidthSpin->value(); }
int SettingsDialog::getCamHeight() const { return camHeightSpin->value(); }
double SettingsDialog::getCamPixelSize() const { return camPixelSpin->value(); }
QString SettingsDialog::getUdpBindIp() const {
    const QString value = udpBindIpEdit->text().trimmed();
    return value.isEmpty() ? QStringLiteral("0.0.0.0") : value;
}
int SettingsDialog::getUdpPort() const { return udpPortSpin->value(); }

// Optical Getters
double SettingsDialog::getWavelength() const { return waveSpin->value(); }
double SettingsDialog::getFocalLength() const { return focalSpin->value(); }
double SettingsDialog::getCameraImagingMagnification() const { return cameraMagnificationSpin->value(); }

// SLM Output Getter
int SettingsDialog::getSlmOutputMode() const { return slmOutputModeCombo->currentIndex(); }

// Algorithm behavior getters
bool SettingsDialog::getAutoRunGsEnabled() const { return autoRunGsCheck->isChecked(); }
bool SettingsDialog::getAutoSendSlmEnabled() const { return autoSendSlmCheck->isChecked(); }
int SettingsDialog::getStartingPhaseMaskMode() const { return startingPhaseMaskCombo->currentData().toInt(); }
int SettingsDialog::getGsComputeBackendMode() const { return computeBackendCombo->currentData().toInt(); }

int SettingsDialog::getOpenClPlatformIndex() const {
    const QString packed = openClDeviceCombo->currentData().toString();
    const QStringList parts = packed.split(':');
    if (parts.size() != 2) {
        return -1;
    }
    bool ok = false;
    const int value = parts.at(0).toInt(&ok);
    return ok ? value : -1;
}

int SettingsDialog::getOpenClDeviceIndex() const {
    const QString packed = openClDeviceCombo->currentData().toString();
    const QStringList parts = packed.split(':');
    if (parts.size() != 2) {
        return -1;
    }
    bool ok = false;
    const int value = parts.at(1).toInt(&ok);
    return ok ? value : -1;
}

int SettingsDialog::getCudaDeviceIndex() const {
    bool ok = false;
    const int value = cudaDeviceCombo->currentData().toInt(&ok);
    return ok ? value : -1;
}

File: python_port/hot_control/ui/settings_dialog.py
from __future__ import annotations

from PySide6.QtCore import Signal
from PySide6.QtWidgets import (
    QCheckBox,
    QComboBox,
    QDialog,
    QDoubleSpinBox,
    QFormLayout,
    QGroupBox,
    QHBoxLayout,
    QLabel,
    QLineEdit,
    QPushButton,
    QSpinBox,
    QTabWidget,
    QVBoxLayout,
    QWidget,
)


class SettingsDialog(QDialog):
    settings_applied = Signal(dict)

    def __init__(self, parent, config, selected_monitor):
        super().__init__(parent)
        self._config = dict(config)
        self._selected_monitor = int(selected_monitor)
        self.setWindowTitle("Hardware Settings")
        self.resize(760, 980)

        root = QVBoxLayout(self)
        tabs = QTabWidget()
        root.addWidget(tabs, 1)

        hardware_tab = QWidget()
        hardware_layout = QVBoxLayout(hardware_tab)
        tabs.addTab(hardware_tab, "Hardware")

        ui_tab = QWidget()
        ui_layout = QVBoxLayout(ui_tab)
        tabs.addTab(ui_tab, "UI Settings")

        slm_group = QGroupBox("SLM Parameters")
        slm_form = QFormLayout(slm_group)
        slm_form.setLabelAlignment(
            slm_form.labelAlignment()
        )
        self.slm_width = QSpinBox()
        self.slm_width.setRange(64, 4096)
        self.slm_width.setValue(int(config.get("slm_width", 1920)))
        self.slm_height = QSpinBox()
        self.slm_height.setRange(64, 4096)
        self.slm_height.setValue(int(config.get("slm_height", 1080)))
        self.slm_px = QDoubleSpinBox()
        self.slm_px.setRange(0.1, 50.0)
        self.slm_px.setDecimals(2)
        self.slm_px.setSuffix(" um")
        self.slm_px.setValue(float(config.get("slm_pixel_size_um", 8.0)))

        self.slm_output_mode = QComboBox()
        self.slm_output_mode.addItem("Direct Screen", "direct")
        self.slm_output_mode.addItem("DLL", "dll")
        out_mode = str(config.get("slm_output_mode", "direct")).lower().strip()
        self.slm_output_mode.setCurrentIndex(1 if out_mode == "dll" else 0)

        self.auto_run_gs = QCheckBox()
        self.auto_run_gs.setChecked(bool(config.get("auto_run_gs", False)))
        self.auto_send_slm = QCheckBox()
        self.auto_send_slm.setChecked(bool(config.get("auto_send_slm", False)))

        self.starting_phase = QComboBox()
        self.starting_phase.addItems(["Checkerboard Pattern", "Binary Grating", "Random Phase"])
        self.starting_phase.setCurrentIndex(int(config.get("gs_starting_phase_mask", 0)))

        self.gs_compute_backend = QComboBox()
        self.gs_compute_backend.addItems(["Auto", "CPU", "OpenCL", "CUDA"])
        backend_idx = int(config.get("gs_compute_backend", 0))
        if backend_idx < 0 or backend_idx > 3:
            backend_idx = 0
        self.gs_compute_backend.setCurrentIndex(backend_idx)

        self.opencl_device = QComboBox()
        self.opencl_device.addItem("[P1:D0] Placeholder OpenCL Device")
        self.opencl_device.setCurrentIndex(0)

        self.cuda_device = QComboBox()
        self.cuda_device.addItem("No CUDA devices detected")
        self.cuda_device.setCurrentIndex(0)

        slm_form.addRow("SLM Width (pixels):", self.slm_width)
        slm_form.addRow("SLM Height (pixels):", self.slm_height)
        slm_form.addRow("SLM Pixel Size:", self.slm_px)
        slm_form.addRow("SLM Output Mode:", self.slm_output_mode)
        slm_form.addRow("Auto-run GS:", self.auto_run_gs)
        slm_form.addRow("Auto-send SLM:", self.auto_send_slm)
        slm_form.addRow("Starting Phase Mask:", self.starting_phase)
        slm_form.addRow("GS Compute Backend:", self.gs_compute_backend)
        slm_form.addRow("OpenCL Device:", self.opencl_device)
        slm_form.addRow("CUDA Device:", self.cuda_device)

        camera_group = QGroupBox("Camera Parameters")
        camera_form = QFormLayout(camera_group)
        self.camera_backend = QComboBox()
        self.camera_backend.addItems(["Qt Native", "OpenCV DirectShow", "UDP Stream (Ethernet)"])
        backend_idx = int(config.get("camera_backend", 0))
        if backend_idx < 0 or backend_idx > 2:
            backend_idx = 0
        self.camera_backend.setCurrentIndex(backend_idx)

        self.cam_width = QSpinBox()
        self.cam_width.setRange(64, 4096)
        self.cam_width.setValue(int(config.get("cam_width", 1920)))
        self.cam_height = QSpinBox()
        self.cam_height.setRange(64, 4096)
        self.cam_height.setValue(int(config.get("cam_height", 1080)))
        self.cam_px = QDoubleSpinBox()
        self.cam_px.setRange(0.1, 50.0)
        self.cam_px.setDecimals(2)
        self.cam_px.setSuffix(" um")
        self.cam_px.setValue(float(config.get("cam_pixel_size_um", 5.0)))
        self.udp_ip = QLineEdit(str(config.get("udp_bind_ip", "0.0.0.0")))
        self.udp_port = QSpinBox()
        self.udp_port.setRange(1, 65535)
        self.udp_port.setValue(int(config.get("udp_port", 9000)))

        camera_form.addRow("Camera Engine:", self.camera_backend)
        camera_form.addRow("Camera Width (px):", self.cam_width)
        camera_form.addRow("Camera Height (px):", self.cam_height)
        camera_form.addRow("Camera Pixel Size:", self.cam_px)
        camera_form.addRow("UDP Bind IP:", self.udp_ip)
        camera_form.addRow("UDP Port:", self.udp_port)

        optical_group = QGroupBox("Optical Setup")
        optical_form = QFormLayout(optical_group)
        self.wave = QDoubleSpinBox()
        self.wave.setRange(100.0, 3000.0)
        self.wave.setDecimals(2)
        self.wave.setSuffix(" nm")
        self.wave.setValue(float(config.get("wavelength_nm", 1064.0)))
        self.focal = QDoubleSpinBox()
        self.focal.setRange(0.1, 2000.0)
        self.focal.setDecimals(3)
        self.focal.setSuffix(" mm")
        self.focal.setValue(float(config.get("focal_length_mm", 100.0)))
        self.cam_mag = QDoubleSpinBox()
        self.cam_mag.setRange(0.01, 100.0)
        self.cam_mag.setDecimals(3)
        self.cam_mag.setValue(float(config.get("camera_imaging_magnification", 1.0)))
        self.monitor = QSpinBox()
        self.monitor.setRange(1, 16)
        self.monitor.setValue(self._selected_monitor)

        optical_form.addRow("Laser Wavelength:", self.wave)
        optical_form.addRow("Fourier Lens Focal Length (physical):", self.focal)
        optical_form.addRow("Camera Imaging Magnification:", self.cam_mag)
        optical_form.addRow("SLM Monitor Number:", self.monitor)

        hardware_layout.addWidget(slm_group)
        hardware_layout.addWidget(camera_group)
        hardware_layout.addWidget(optical_group)
        hardware_layout.addWidget(QLabel("Monitor target is selected from Tools > Select Monitor"))
        hardware_layout.addStretch()

        self.ui_dark_mode = QCheckBox("Dark Theme")
        self.ui_dark_mode.setChecked(bool(config.get("ui_dark_mode", True)))
        self.ui_show_overlay = QCheckBox("Enable Overlay Target by Default")
        self.ui_show_overlay.setChecked(bool(config.get("ui_overlay_default", False)))
        ui_layout.addWidget(self.ui_dark_mode)
        ui_layout.addWidget(self.ui_show_overlay)
        ui_layout.addStretch()

        btn_row = QHBoxLayout()
        btn_row.addStretch()
        self.save_btn = QPushButton("Save")
        self.cancel_btn = QPushButton("Cancel")
        self.apply_btn = QPushButton("Apply")
        self.save_btn.clicked.connect(self._on_save)
        self.cancel_btn.clicked.connect(self.reject)
        self.apply_btn.clicked.connect(self._on_apply)
        btn_row.addWidget(self.save_btn)
        btn_row.addWidget(self.cancel_btn)
        btn_row.addWidget(self.apply_btn)
        root.addLayout(btn_row)

    def _collect_payload(self):
        out = dict(self._config)
        out["slm_width"] = int(self.slm_width.value())
        out["slm_height"] = int(self.slm_height.value())
        out["slm_pixel_size_um"] = float(self.slm_px.value())
        out["slm_output_mode"] = str(self.slm_output_mode.currentData())
        out["auto_run_gs"] = bool(self.auto_run_gs.isChecked())
        out["auto_send_slm"] = bool(self.auto_send_slm.isChecked())
        out["gs_starting_phase_mask"] = int(self.starting_phase.currentIndex())
        out["gs_compute_backend"] = int(self.gs_compute_backend.currentIndex())
        out["cam_width"] = int(self.cam_width.value())
        out["cam_height"] = int(self.cam_height.value())
        out["cam_pixel_size_um"] = float(self.cam_px.value())
        out["camera_imaging_magnification"] = float(self.cam_mag.value())
        out["wavelength_nm"] = float(self.wave.value())
        out["focal_length_mm"] = float(self.focal.value())
        out["camera_backend"] = int(self.camera_backend.currentIndex())
        out["udp_bind_ip"] = self.udp_ip.text().strip() or "0.0.0.0"
        out["udp_port"] = int(self.udp_port.value())
        out["ui_dark_mode"] = bool(self.ui_dark_mode.isChecked())
        out["ui_overlay_default"] = bool(self.ui_show_overlay.isChecked())
        return {"config": out, "selected_monitor": int(self.monitor.value())}

    def _on_apply(self):
        self.settings_applied.emit(self._collect_payload())

    def _on_save(self):
        self.accept()

    def result_payload(self):
        return self._collect_payload()
